### PR TITLE
feat(permissions): make origin optional

### PR DESCRIPTION
**Pull request by pavelfeldman**

### CODE REVIEW

File: docs/api.md [hunk 1]
```diff
@@ -544,9 +544,9 @@ To disable authentication, pass `null`.
 - `offline` <[boolean]> Whether to emulate network being offline for the browser context.
 - returns: <[Promise]>
 
-#### browserContext.setPermissions(origin, permissions[])
-- `origin` <[string]> The [origin] to grant permissions to, e.g. "https://example.com".
-- `permissions` <[Array]<[string]>> An array of permissions to grant. All permissions that are not listed here will be automatically denied. Permissions can be one of the following values:
+#### browserContext.grantPermissions(permissions[], options)
```
Comment (dgozman) [inline docs/api.md]: Should be `(permissions[, options])`

File: docs/api.md [hunk 2]
```diff
@@ -381,7 +381,7 @@ Clears all permission overrides for the browser context.
 
 ```js
 const context = await browser.newContext();
-context.setPermissions('https://example.com', ['clipboard-read']);
+context.grantPermissions(['clipboard-read']);
```
Comment (dgozman) [inline docs/api.md]: `await context.grantPermissions(...)`

File: docs/api.md [hunk 3]
```diff
@@ -381,7 +381,7 @@ Clears all permission overrides for the browser context.
...
 // do stuff ..
 context.clearPermissions();
```
Comment (dgozman) [inline docs/api.md]: await here as well.

File: package.json [hunk 4]
```diff
@@ -46,6 +46,7 @@
     "extract-zip": "^1.6.6",
     "https-proxy-agent": "^3.0.0",
     "jpeg-js": "^0.3.6",
+    "playwright": "^0.11.1-next.1583993157193",
```
Comment (dgozman) [inline package.json]: Stray change?

File: src/browserContext.ts [hunk 5]
```diff
@@ -103,6 +103,19 @@ export abstract class BrowserContextBase extends platform.EventEmitter implement
...
+    const origin = options && options.origin || '*';
+    const existing = new Set(this._permissions.get(origin) || []);
+    permissions.forEach(p => existing.add(p));
+    const list = [...existing.values()];
+    this._permissions.set(origin, list);
+    await this._doGrantPermissions(origin, list);
+  }
+
+  async clearPermissions() {
+    this._permissions.clear();
```
Comment (dgozman) [inline src/browserContext.ts]: It is confusing that we don't call `_doClearPermissions` here.

File: src/chromium/crBrowser.ts [hunk 6]
```diff
@@ -327,10 +327,11 @@ export class CRBrowserContext extends BrowserContextBase {
         throw new Error('Unknown permission: ' + permission);
       return protocolPermission;
     });
-    await this._browser._session.send('Browser.grantPermissions', { origin, browserContextId: this._browserContextId || undefined, permissions: filtered });
+    await this._browser._session.send('Browser.grantPermissions', { origin: origin === '*' ? undefined : origin, browserContextId: this._browserContextId || undefined, permissions: filtered });
   }
 
   async clearPermissions() {
+    super.clearPermissions();
```
Comment (dgozman) [inline src/chromium/crBrowser.ts]: Please await this, since it's an async function.

File: src/webkit/wkBrowser.ts [hunk 7]
```diff
@@ -190,8 +190,8 @@ export class WKBrowserContext extends BrowserContextBase {
       await this._browser._browserSession.send('Playwright.setIgnoreCertificateErrors', { browserContextId: this._browserContextId, ignore: true });
     if (this._options.locale)
       await this._browser._browserSession.send('Playwright.setLanguages', { browserContextId: this._browserContextId, languages: [this._options.locale] });
-    const entries = Object.entries(this._options.permissions || {});
-    await Promise.all(entries.map(entry => this.setPermissions(entry[0], entry[1])));
+    if (this._options.permissions)
+      this.grantPermissions(this._options.permissions);
```
Comment (dgozman) [inline src/webkit/wkBrowser.ts]: await it

File: test/permissions.spec.js [hunk 8]
```diff
@@ -38,23 +38,55 @@ module.exports.describe = function({testRunner, expect, WEBKIT}) {
...
       await page.goto(server.EMPTY_PAGE);
-      await context.setPermissions(server.EMPTY_PAGE, []);
+      await context.grantPermissions([], { origin: server.EMPTY_PAGE });
       expect(await getPermission(page, 'geolocation')).toBe('denied');
     });
     it('should fail when bad permission is given', async({page, server, context}) => {
       await page.goto(server.EMPTY_PAGE);
       let error = {};
-      await context.setPermissions(server.EMPTY_PAGE, ['foo']).catch(e => error = e);
+      await context.grantPermissions(['foo']).catch(e => error = e, { origin: server.EMPTY_PAGE });
```
Comment (dgozman) [inline test/permissions.spec.js]: This is some funky syntax 😄

File: test/permissions.spec.js [hunk 9]
```diff
@@ -71,9 +103,9 @@ module.exports.describe = function({testRunner, expect, WEBKIT}) {
         });
       });
       expect(await page.evaluate(() => window['events'])).toEqual(['prompt']);
-      await context.setPermissions(server.EMPTY_PAGE, []);
+      await context.grantPermissions([], { origin: server.EMPTY_PAGE });
```
Comment (dgozman) [inline test/permissions.spec.js]: `server.EMPTY_PAGE` is not quite an origin. Do we support urls? I don't see where we convert url to an origin.

**Review by dgozman — APPROVED** [review]

File: package.json (hunk 4 above)
Comment (pavelfeldman) [reply package.json]: Very

File: docs/api.md (hunk 3 above)
Comment (pavelfeldman) [reply docs/api.md]: Done as well

File: docs/api.md (hunk 2 above)
Comment (pavelfeldman) [reply docs/api.md]: Done

File: docs/api.md (hunk 1 above)
Comment (pavelfeldman) [reply docs/api.md]: Done

File: src/browserContext.ts (hunk 5 above)
Comment (pavelfeldman) [reply src/browserContext.ts]: I thought the same way. Fixing.

File: src/chromium/crBrowser.ts (hunk 6 above)
Comment (pavelfeldman) [reply src/chromium/crBrowser.ts]: Done

File: src/webkit/wkBrowser.ts (hunk 7 above)
Comment (pavelfeldman) [reply src/webkit/wkBrowser.ts]: Done

File: test/permissions.spec.js (hunk 8 above)
Comment (pavelfeldman) [reply test/permissions.spec.js]: ooo

File: test/permissions.spec.js (hunk 9 above)
Comment (pavelfeldman) [reply test/permissions.spec.js]: I think it happens inside the browsers :/

**Pull request merged**